### PR TITLE
switch from rasterio/gdal to pyproj

### PR DESCRIPTION
**Pull request by vincentsarago**
as seen in #56 and #57 we are constantly fighting with change in rasterio/gdal/proj. The main problem is that we used rasterio (which is a raster library) to access Proj C libs via GDAL. It worked ok, and we choose this to avoid mixing PROJ version used in the code base (for a long time rasterio wheels where using PROJ 6 while pyproj was already using PROJ>=7). 

This PR replace any rasterio calls by pyproj.

⚠️ this is a highly breaking change ⚠️

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -13,7 +13,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.6, 3.7, 3.8, 3.9]
+        python-version: [3.7, 3.8, 3.9]
```
Comment (vincentsarago) [inline .github/workflows/ci.yml:16]: pyproj only supports >=3.7

File: morecantile/commons.py [hunk 2]
```diff
@@ -2,10 +2,9 @@
 
 from collections import OrderedDict, namedtuple
 
-from rasterio.coords import BoundingBox  # noqa
-
 _Tile = namedtuple("_Tile", ["x", "y", "z"])
 _Coords = namedtuple("_Coords", ["x", "y"])
+BoundingBox = namedtuple("BoundingBox", ("left", "bottom", "right", "top"))
```
Comment (vincentsarago) [inline morecantile/commons.py:7]: maybe it's also time to move away from nametuple 

cc @kylebarron

File: morecantile/data/WorldCRS84Quad.json [hunk 3]
```diff
@@ -1,255 +0,0 @@
-{
-    "type": "TileMatrixSetType",
-    "title": "CRS84 for the World",
-    "identifier": "WorldCRS84Quad",
-    "boundingBox": {
-        "type": "BoundingBoxType",
-        "crs": "http://www.opengis.net/def/crs/OGC/1.3/CRS84",
```
Comment (vincentsarago) [inline morecantile/data/WorldCRS84Quad.json:7]: pyproj doesn't recognize this CRS 🤷‍♂️

File: morecantile/defaults.py [hunk 4]
```diff
@@ -56,4 +56,4 @@ def register(
         return TileMatrixSets({**self.tms, **new_tms})
 
 
-tms = TileMatrixSets(deepcopy(default_tms))  # noqa
+tms = TileMatrixSets(copy(default_tms))  # noqa
```
Comment (vincentsarago) [inline morecantile/defaults.py:59]: deepcopy complained about the pyproj Transformer. Used `copy` and it worked 🤷‍♂️

File: morecantile/models.py [hunk 5]
```diff
@@ -40,7 +39,11 @@ def validate(cls, value: Union[CRS, AnyHttpUrl]):
         """Validate CRS."""
         # If input is a string we tranlate it to CRS
         if not isinstance(value, CRS):
-            return CRS.from_user_input(value)
+            try:
+                return CRS.from_user_input(value)
+            except CRSError:
+                return CRS.from_epsg(value.split("/")[-1])
```
Comment (vincentsarago) [inline morecantile/models.py:45]: some URI are not valid 🤷‍♂️

Comment (vincentsarago) [reply morecantile/models.py:45]: I should check with the uri has `EPSG` in it

File: morecantile/models.py [hunk 6]
```diff
@@ -121,13 +124,25 @@ class TileMatrixSet(BaseModel):
     wellKnownScaleSet: Optional[AnyHttpUrl] = None
     boundingBox: Optional[TMSBoundingBox]
     tileMatrix: List[TileMatrix]
+    _to_wgs84: Transformer = PrivateAttr()
+    _from_wgs84: Transformer = PrivateAttr()
```
Comment (vincentsarago) [inline morecantile/models.py:128]: we initialize the transformers once!

File: tests/test_morecantile.py [hunk 7]
```diff
@@ -247,47 +238,23 @@ def test_xy_null_island():
...
-def test_xy_south_pole_gdal3():
-    """Return -inf for y at South Pole"""
+    Note: mercantile returns (0.0, inf)
+    """
     tms = morecantile.tms.get("WebMercatorQuad")
     with pytest.warns(PointOutsideTMSBounds):
         xy = tms.xy(0.0, -90)
         assert xy.x == float("inf")
-        assert xy.y == float("-inf")
+        assert xy.y == float("inf")
```
Comment (vincentsarago) [inline tests/test_morecantile.py:250]: inf or -inf 🤷‍♂️

File: morecantile/models.py [hunk 8]
```diff
@@ -147,6 +170,19 @@ def crs(self) -> CRS:
...
+        """Return rasterio CRS."""
+        if not rasterio:
+            raise ModuleNotFoundError(
+                "Rasterio has to be installed to use `rasterio_crs` method."
+            )
+
+        if GDALVersion.runtime().major < 3:
+            return rasterio.crs.CRS.from_wkt(self.crs.to_wkt(WktVersion.WKT1_GDAL))
+        else:
+            return rasterio.crs.CRS.from_wkt(self.crs.to_wkt())
```
Comment (vincentsarago) [inline morecantile/models.py:184]: from https://pyproj4.github.io/pyproj/stable/crs_compatibility.html#rasterio

File: morecantile/models.py [hunk 9]
```diff
@@ -147,6 +170,19 @@ def crs(self) -> CRS:
         """Fetch CRS from epsg"""
         return self.supportedCRS
 
+    @property
+    def rasterio_crs(self) -> CRS:
```
Comment (vincentsarago) [inline morecantile/models.py:174]: The return type is not the good one here

File: morecantile/data/CanadianNAD83_LCC.json [hunk 10]
```diff
@@ -4,7 +4,7 @@
     "identifier": "CanadianNAD83_LCC",
     "boundingBox": {
         "type": "BoundingBoxType",
-        "crs": "http://www.opengis.net/def/crs/EPSG/0/3978",
+        "crs": "urn:ogc:def:crs:EPSG::3978",
```
Comment (vincentsarago) [inline morecantile/data/CanadianNAD83_LCC.json:7]: pyproj seems to prefers those

File: morecantile/models.py [hunk 11]
```diff
@@ -149,12 +149,21 @@ def __init__(self, **data):
...
-            WGS84_CRS, self.supportedCRS, always_xy=True
-        )
+        try:
+            self._to_wgs84 = Transformer.from_crs(
+                self.supportedCRS, WGS84_CRS, always_xy=True
+            )
+            self._from_wgs84 = Transformer.from_crs(
+                WGS84_CRS, self.supportedCRS, always_xy=True
+            )
+        except ProjError:
```
Comment (vincentsarago) [inline morecantile/models.py:159]: Sometime (e.g working with non earth projection) we cannot create Transformer with WGS84 srs

File: morecantile/models.py [hunk 12]
```diff
@@ -42,11 +49,12 @@ def __get_validators__(cls):
         yield cls.validate
 
     @classmethod
-    def validate(cls, value: Union[CRS, AnyHttpUrl]):
+    def validate(cls, value: Union[CRS, AnyHttpUrl]) -> CRS:
```
Comment (vincentsarago) [inline morecantile/models.py:52]: `AnyHttpUrl` -> `str`: `urn:ogc:def:crs:EPSG::2193` is not an url

File: morecantile/commons.py (hunk 2 above)
Comment (kylebarron) [reply morecantile/commons.py:7]: If you stick with `namedtuple`, I like the new class-like syntax of

```py
from typing import NamedTuple

class BoundingBox(NamedTuple):
	left: int
	bottom: int
	right: int
	top: int
```

But that's a personal preference 🤷‍♂️

**Pull request merged**